### PR TITLE
[SYCL] Diagnose local accessor use in single_task or parallel_for(range)

### PR DESCRIPTION
**Pull request by mmoadeli**
* According to [local accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:accessor.local) of the SYCL specification, a local accessor must not be used in a SYCL kernel function that is invoked via single_task or via the simple form of parallel_for that takes a range parameter.
* Add test.

### CODE REVIEW

**Comment by mmoadeli** [comment]
/Verify with https://github.com/intel/llvm-test-suite/pull/1646

**Comment by mmoadeli** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1646

File: sycl/include/sycl/handler.hpp [hunk 1]
```diff
@@ -354,6 +354,31 @@ class __SYCL_EXPORT handler {
                                 PI_ERROR_INVALID_OPERATION);
   }
 
+  /// According to section 4.7.6.11. of the SYCL specification, a local accessor
+  /// must not be used in a SYCL kernel function that is invoked via single_task
+  /// or via the simple form of parallel_for that takes a range parameter.
+  template <typename KernelName, typename KernelType>
+  void throwOnLocalAccessorMisuse() {
```
Comment (steffenlarsen) [inline sycl/include/sycl/handler.hpp:361]:
```suggestion
  void throwOnLocalAccessorMisuse() const {
```

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by mmoadeli** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1646

**Comment by mmoadeli** [comment]
@bader there are a number of places in corresponding PR at llvm-test-suite. I'll have them sorted out by tomorrow. That will make both ready to merge.

**Comment by bader** [comment]
> @bader there are a number of places in corresponding PR at llvm-test-suite. I'll have them sorted out by tomorrow. That will make both ready to merge.

👍 
Let @intel/llvm-gatekeepers know when it's ready.

**Comment by mmoadeli** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1646

**Comment by steffenlarsen** [comment]
Windows L0 Failed Tests (1):
SYCL :: USM/memcpy2d.cpp
Unrelated. Has been temporarily disabled.

**Comment by bader** [comment]
@mmoadeli, I had to revert this patch because it introduced a lot of failures to both pre- and post-commit validation. Please, fix all these issues before re-applying the patch again.

**Pull request merged**